storage/cleanup.go

```diff
@@ -132,38 +131,48 @@ func (m *cleanupManager) cleanupFilesetFiles(earliestToRetain time.Time) error {
 // writes and future writes, we need to shift flush time range by block size as the
 // time range for commit log files we need to check.
 func (m *cleanupManager) commitLogTimeRange(t time.Time) (time.Time, time.Time) {
-	flushStart, flushEnd := m.fm.FlushTimeStart(t), m.fm.FlushTimeEnd(t)
-	return flushStart.Add(-m.blockSize), flushEnd.Add(-m.blockSize)
+	var (
+		ropts      = m.opts.CommitLogOptions().RetentionOptions()
+		blockSize  = ropts.BlockSize()
+		flushStart = retention.FlushTimeStart(ropts, t)
+		flushEnd   = retention.FlushTimeEnd(ropts, t)
+	)
+	return flushStart.Add(-blockSize), flushEnd.Add(-blockSize)
 }
 
 // commitLogTimes returns the earliest time before which the commit logs are expired,
 // as well as a list of times we need to clean up commit log files for.
 func (m *cleanupManager) commitLogTimes(t time.Time) (time.Time, []time.Time) {
-	earliest, latest := m.commitLogTimeRange(t)
-
-	// TODO(xichen): preallocate the slice here
-	var commitLogTimes []time.Time
-	for commitLogTime := latest; !commitLogTime.Before(earliest); commitLogTime = commitLogTime.Add(-m.blockSize) {
-		hasFlushedAll := true
-		leftBlockStart := commitLogTime.Add(-m.blockSize)
-		rightBlockStart := commitLogTime.Add(m.blockSize)
-		for blockStart := leftBlockStart; !blockStart.After(rightBlockStart); blockStart = blockStart.Add(m.blockSize) {
-			if m.fm.NeedsFlush(blockStart) {
-				hasFlushedAll = false
-				break
-			}
-		}
-		if hasFlushedAll {
-			commitLogTimes = append(commitLogTimes, commitLogTime)
+	var (
+		ropts            = m.opts.CommitLogOptions().RetentionOptions()
+		blockSize        = ropts.BlockSize()
+		earliest, latest = m.commitLogTimeRange(t)
+	)
+	// NB(prateek): this logic of polling the namespaces across the commit log's entire
+	// retention history could get expensive if commit logs are retained for long periods.
+	// e.g. if we retain them for 40 days, with a block 2 hours; then every time
+	// we try to flush we are going to be polling each namespace, for each shard, for 480
+	// distinct blockstarts. Say we have 2 namespaces, each with 8192 shards, that's ~10M map lookups.
+	// If we cared about 100% correctness, we would optimize this by retaining a smarter data
+	// structure (e.g. interval tree), but for our use-case, it's safe to assume that commit logs
+	// are only retained for a period of 1-2 days (at most), after we which we'd live we with the
+	// data loss.
+
+	times := make([]time.Time, 0, numIntervals(earliest, latest, blockSize))
+	for t := latest; !t.Before(earliest); t = t.Add(-blockSize) {
+		leftBlockStart, rightBlockStart := t.Add(-blockSize), t.Add(blockSize)
+		if anyPendingFlushes := m.fm.NeedsFlush(leftBlockStart, rightBlockStart); !anyPendingFlushes {
```

Comment:
Yep, NeedsFlush is inclusive on both ends: https://github.com/m3db/m3db/blob/c93ecb1dc6806088f616b5ff8cf76960e9acba90/storage/types.go#L207-L209